File: concurrencyCounter.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <stdexcept>

// Global synchronization primitives
std::mutex mtx;                 // Mutex lock for critical section
std::condition_variable cv;     // Condition variable for thread synchronization
bool countUpFinished = false;   // Flag to indicate when counting is finished

// Function to count up from 0 to 20
void countUp() {
    try {
        for (int i = 0; i <= 20; ++i) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Simulates work being done to demonstrate synchronization concept
            std::cout << "Counting up: " << i << std::endl;
        }
        {
            std::lock_guard<std::mutex> lock(mtx);  // Lock mutex to update the flag
            countUpFinished = true;
        }
        cv.notify_one();    // Notify the waiting thread that counting up is finished
    } catch (const std::exception& e) {
        std::cerr << "Exception in countUp: " << e.what() << std::endl; // Handle exceptions
    }
}

// Function to count down from 20 to 0
void countDown() {
    try {
        std::unique_lock<std::mutex> lock(mtx);         //Acquire mutex lock
        cv.wait(lock, [] { return countUpFinished;});   // Waits until countUp is finished
        for (int i = 20; i >= 0; --i) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));    // Simulates work being done to demonstrate synchronization concept
            std::cout << "Counting down: " << i << std::endl;
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in countDown: " << e.what() << std::endl;   // Hnadle exceptions
    }
}

int main() {
    try {
        // Create and start threads for counting up and down
        std::thread t1(countUp);
        std::thread t2(countDown);

        // Wait for both threads to finish
        t1.join();
        t2.join();
    } catch (const std::exception& e) {
        std::cerr << "Exception in main: " << e.what() << std::endl;    // Handle exceptions
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: concurrencyCounter.java
import java.util.logging.Level;
import java.util.logging.Logger;

public class concurrencyCounter {

    private static final Logger logger = Logger.getLogger(concurrencyCounter.class.getName());
    private static final int SLEEP_DURATION = 100;
    private static final int COUNT_UP_LIMIT = 20;
    private static final int COUNT_DOWN_START = 20;
    private static final int COUNT_DOWN_END = 0;

    public static void main(String[] args) {
        // Create and name threads
        Thread countUp = new Thread(new CountUp(), "CountUpThread");
        Thread countDown = new Thread(new CountDown(), "CountDownThread");

        // Start the countUp thread
        countUp.start();

        // Wait for the countUp thread to finish
        try {
            countUp.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt(); // Reset the interrupt status
            logger.log(Level.SEVERE, "Main thread was interrupted while waiting for countUp.", e);
        }

        // Start the countDown thread
        countDown.start();

        // Wait for the countDown thread to finish
        try {
            countDown.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt(); // Reset the interrupt status
            logger.log(Level.SEVERE, "Main thread was interrupted while waiting for countDown.", e);
        }
    }

    // Runnable class that counts up from 1 to 20
    static class CountUp implements Runnable {
        @Override
        public void run() {
            try {
                for (int i = 1; i <= COUNT_UP_LIMIT; i++) { // Count up from 1 to 20
                    System.out.println("Count Up: " + i);
                    Thread.sleep(SLEEP_DURATION); // Sleep to simulate work to demonstrate concurrency
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); // Handle the interruption of the CountUp thread
                logger.log(Level.WARNING, "CountUp thread was interrupted.", e);
            }
        }
    }

    // Runnable class that counts down from 20 to 0
    static class CountDown implements Runnable {
        @Override
        public void run() {
            try {
                for (int i = COUNT_DOWN_START; i >= COUNT_DOWN_END; i--) { // Count down from 20 to 0
                    System.out.println("Count Down: " + i);
                    Thread.sleep(SLEEP_DURATION); // Sleep to simulate work to demonstrate concurrency
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); // Handle the interruption of the CountDown thread
                logger.log(Level.WARNING, "CountDown thread was interrupted.", e);
            }
        }
    }
}
